tests/database/test_batchable_db.py

```diff
@@ -0,0 +1,131 @@
+import pytest
+
+from eth_utils import ValidationError
+
+from eth.db.backends.memory import MemoryDB
+from eth.db.batch import BatchableDB
+
+
+@pytest.fixture
+def base_db():
+    return MemoryDB()
+
+
+@pytest.fixture
+def batch_db(base_db):
+    return BatchableDB(base_db)
+
+
+def test_batchable_db_with_set_and_get(base_db, batch_db):
+    with batch_db.write_batch():
+        batch_db.set(b'key-1', b'value-1')
+        batch_db.set(b'key-2', b'value-2')
+        assert batch_db.get(b'key-1') == b'value-1'
+        assert batch_db.get(b'key-2') == b'value-2'
+
+        # keys should not yet be set in base db.
+        assert b'key-1' not in base_db
+        assert b'key-2' not in base_db
+
+    assert base_db.get(b'key-1') == b'value-1'
+    assert base_db.get(b'key-2') == b'value-2'
+
+
+def test_batchable_db_with_set_and_get_unbatched(base_db, batch_db):
+    batch_db.set(b'key-1', b'value-1')
+    assert batch_db.get(b'key-1') == b'value-1'
+    batch_db.set(b'key-2', b'value-2')
+    assert batch_db.get(b'key-2') == b'value-2'
+
+    # keys should be immediately set in base db.
+    assert base_db.get(b'key-1') == b'value-1'
+    assert base_db.get(b'key-2') == b'value-2'
+
+
+def test_batchable_db_cannot_recursively_batch(base_db, batch_db):
+    with pytest.raises(ValidationError):
```

Comment:
To accurately know where the exception comes from, wouldn't it be cleaner to have this inside the first `with batch_db.write_batch()` and above the inner `with batch_db.write_batch`?

The way this is currently written, the outer call to `write_batch` could be throwing the `ValidationError` and the test would pass.

Reply:
Yeah, I got stuck on this.

If `pytest.raises` is inside the first context, then the it catches the exception, and the outer writes will succeed. We know the outer one works from other tests. I wanted to test that the outer writes will be reverted. But I guess we know that works from other tests, too.

I guess I'll move it inside.